# Analysis of App Profiles for Profit
`
        This project is for finding app profiles based on data, that will attract more users. Considering that we want to build apps that are free to download and install, the main revenue source consists of in-app ads. Hence, the amount of users downloading the app and using it, is what matters in terms of profit. Hence, for making profitable apps, we need to analyse data for looking the category of apps in which most users are interested in.`

`**Goal**: Analyze data to help developers understand what type of apps are likely to attract more users. Using the data based on various factors such as user ratings, suggest the category of apps that may turn out to be more profitable

## Opening and reading the data set
   First we open the two .csv data set files and read them. We use the `reader()` function from the csv module. We store these read files in variables in form of 2D list or list of lists.
        
 - Play Store Data set link: [play_data]("https://www.kaggle.com/lava18/google-play-store-apps")
 - App Store Data set link: [app_data]("https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps")
        

In [1]:
from csv import reader

### App Store data sample ###
opened_file = open("AppleStore.csv")
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

### Play Store data sample ###
opened_file = open("googleplaystore.csv")
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]


We'll be using a function called `explore_data()`, which will help us to print data more clearly and show number of rows and columns, making it more readable.

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') 

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
print(android_header)  
print("\n")
print(explore_data(android, 0, 4, True)) 

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10841
Number of columns: 13
None


_The above was the data from Play store. Let's take a look at the App store data._

In [3]:
print(ios_header)
print("\n")
print(explore_data(ios, 0, 4, True))

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7197
Number of columns: 16
None


The data sets have 7197 iOS app data and 10841 android app data.
By observing both the data sets, we can point out some of the columns that might be useful for our analysis, by looking at the headers.
   
   For example; In Play Store data, the columns 'Rating', 'Review', 'Type', 'Content Rating' seem useful.
   
   And, in App Store, the columns 'user-rating', 'price', 'prime_genre', 'rating_count_tot' seem useful.
   
   ## Deleting Incorrect Data
   The Google Play data set has a dedicated [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), and we can see that [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472. Let's print this row and compare it against the header and another row that is correct.

In [4]:
del android[10472]           #Do not execute this more than once!!!
print("Deleted succesfully")

Deleted succesfully


We found that the entry number 10472 ('Life Made WI-Fi Touchscreen Photo Frame') had rating of 19, which was wrong as the max app rating for an app in play store is 5. Hence, we deleted that.

## Removing Duplicate data examples
   Looking at the data set for some time and reading the [discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion) section, we can observe that there are some duplicate entries of some apps.
   For example; let's take a look at number of entries of Instagram

In [5]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


By seperating unique and duplicated apps, we can observe that there are 1181 cases where an app occurs more than once.

In [6]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])

Number of duplicate apps 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


We don't want an app more than once when we analyse data, so we need to remove the duplicate apps and keep just a sigle entry for a app. Now, we could remove duplicate rows randomly, but there is a better way.

If we examine the rows printed for all entries of Instagram, the main difference happens on the 4th column i.e 4th position in the row. It is the number of reviews. The different numbers imply that the data was collected at different times.

We can use this criterion for removing duplicates. The higher the number of reviews, the more recent the data should be. So, we'll keep the row which has the highest number of reviews.

One of the ways to remove the duplicates, would be to use a dictionary. Each element of a dictionary has a unique value, hence there can't be any duplicates. So the way we execute this is;
 - Create a dictionary, where the keys would be the app names, and the values of the respective keys will be the the highest number of reviews for each app.
 - Using the dictionary, create a new data set, which will have only one entry per app ( and for each app, we'll only select the entry with highest number of reviews).
  Let's build the dictionary now.

In [7]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews :
        reviews_max[name] = n_reviews
    elif name not in reviews_max :
        reviews_max[name] = n_reviews
        

In the previous code cell, we found that there were 1181 duplicate apps. Since we removed all the duplicates, we should get difference between the lengths of original entries and new entries as 9659.

In [8]:
print('Expected length:', len(android) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


Now, we use the dictionary created to remove duplicate rows.
We do this by creating two empty lists; one for our cleaned data set and the other for storing app names. We will use these app names to make sure we don't have duplicates of same number of reviews. We execute this process in two steps;
 - We start by initializing two empty lists, android_clean and already_added.
 - We loop through the android data set, and for every iteration:
    - We isolate the name of the app and the number of reviews.
    - We add the current row (app) to the android_clean list, and the app name (name) to the already_added list if:
       - The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and
       - The name of the app is not already in the already_added list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [9]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)

Let's see how the cleaned data looks like, and verify if there are exactly 9659 rows.

In [10]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


## Removing Non-English Apps
We use only English for the apps we will be developing, after we finish analysing data. Hence, we would like to analyze only the apps that are directed towards an English-speaking audience. However, after exploring the data long enough, we can see that both data sets have apps with names that suggest they are not directed toward an English speaking audience.

For example;

In [11]:
print(ios[813][1])
print(ios[6731][1])
print('\n')
print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


中国語 AQリスニング
لعبة تقدر تربح DZ


Since, we are not interested in keeping such apps, we'll remove them. One way to go about this is to remove each app with a name containing a symbol that is not commonly used in English text.

All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

So, if an app contains a character is greater than 127, then that character is probably non-English. So, we'll build a function, that uses built-in ord() function,  to find out the corresponding encoding number of each character.

In [12]:
def is_english(string):
    for char in string:
        if ord(char) > 127:
            return False
    return True    

Let's test our function with some app names, to check if they detect an app as English or non-English.

In [13]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


Our function works fine, but there emojis and some special symbols(™, — (em dash), – (en dash), etc.), that are detected as non-English, since they fall out of the ASCII range of 0 - 127. 

Because of this, we'll be losing a lot of useful data, since many English apps are incorrectly labeled as non-English.

One of the solution is, that we'll only label apps as non-English, if there are more than three characters that fall out of the ASCII Range. So we will change the is_English() function, and test again.

In [14]:
def is_english(string):
    count = 0
    for char in string:
        if ord(char) > 127:
            count += 1
        if count > 3:
            return False
    return True 

Lets test this modified is_english() function with some examples;

In [15]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Now it works good. This is_english() function is not perfect, but it should be fairly effective. Hence, we will use this further now.

In [16]:
android_clean2 = []
ios_clean = []

for app in android_clean:
    if is_english(app[0]) == True:
        android_clean2.append(app)
        
for app in ios:
    if is_english(app[1]) == True:
        ios_clean.append(app)      

Most of the non-English apps should be filtered now. Let's take a look at few rows of our filtered data, and check how many rows are left in each.

In [17]:
explore_data(android_clean2, 0, 3, True)
print('\n')
explore_data(ios_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

So now, we have 9614 android app data, and 7197 iOS app data.
Next we will isolate free apps from paid apps, as we are trying analyse data of free apps.
## Isolating the Free Apps
As mentioned in the intro, we are analyzing data with the aim to develop apps that are free of cost. Hence, we analyze data that has only free of cost apps in it. The current data we have has both paid and free apps. Now, we will isolate free apps from data set.

In [18]:
android_final = []
ios_final = []

for app in android_clean2:
    price = app[6]    
    if price == 'Free':
        android_final.append(app)
        
for app in ios_clean:
    price = float(app[4])
    if price == 0.0:
        ios_final.append(app)

Now the new data should only have free apps. Let's see the number of rows.

In [19]:
print(len(android_final))
print(len(ios_final))

8863
3222


So we are left with 8863 android app data and 3222 iOS app data, which should be enough for our analysis.
## Common Genres
We are aiming to develop a free app and our revenue is dependent on the number of people using our appps.
Hence, to minimize risk and overhead, our validation strategy for an app data is comprised of three steps:
   1. Build a minimal Android version of the app, and add it to Google Play.
   2. If the app has a good response from users, we develop it further.
   3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by trying to find out the most common genres for each market. For this, we'll build a frequency table for the prime_genre column of the App Store data set, and the Genres and Category columns of the Google Play data set.

We will build two functions for the prime_genre column of App Store data, and another for Genres and Category columns of Google Play data. The two functions are to analyze frequency tables.
  - One function to generate frequency tables that show percentages.
  - Another function to display the percentages in descending order.

In [20]:

def freq_table(dataset, index):
    table = {}
    total = 0
    for data in dataset:
        total += 1
        genre = data[index]
        if genre in table:
            table[genre] += 1
        else:
            table[genre] = 1
            
    table_percentage = {}        
    for key in table:
        table_percentage[key] = (table[key] / total) * 100
        
    return table_percentage

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

We start by examining the frequency table for the prime_genre column of the App Store data set.

In [21]:
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


It looks like the Gaming genre is the most popular with 58.16%, in the free English apps category. After games, the entertainment genre is the runner-up with almost 8%. Then there is Photo & video (almost 5%), Education(3.66%), Social Networking(3.28%).... and so on.

The general impression is that, most apps on App Store are designed for entertainment purposes, in the category of free English based apps. And, the apps designed for practical purposes(education, shopping, utilities,etc) are rare. 

If an app profile is to be recommended for the App Store market based on this data, we may say that a unique new game or some entertainment app may turn out to be profitable. But this data is the apps uploaded by other developers, which means this just tells what types of app are supplied more to the App Store. This doesn't imply the users' demand, but the number of apps put on App Store from a specific genre. Hence, the demand might not be the same as the offer.

Let's continue and analyse genre and catgory columns of the Google Play Store data.

In [22]:
display_table(android_final, 1) # Category

FAMILY : 18.898792733837304
GAME : 9.725826469592688
TOOLS : 8.462146000225657
BUSINESS : 4.592124562789123
LIFESTYLE : 3.9038700214374367
PRODUCTIVITY : 3.8925871601038025
FINANCE : 3.7007785174320205
MEDICAL : 3.5315355974275078
SPORTS : 3.396141261423897
PERSONALIZATION : 3.317161232088458
COMMUNICATION : 3.2381812027530184
HEALTH_AND_FITNESS : 3.0802211440821394
PHOTOGRAPHY : 2.944826808078529
NEWS_AND_MAGAZINES : 2.798149610741284
SOCIAL : 2.6627552747376737
TRAVEL_AND_LOCAL : 2.335552296062281
SHOPPING : 2.245289405393208
BOOKS_AND_REFERENCE : 2.1437436533904997
DATING : 1.8616721200496444
VIDEO_PLAYERS : 1.7939749520478394
MAPS_AND_NAVIGATION : 1.399074805370642
FOOD_AND_DRINK : 1.241114746699763
EDUCATION : 1.1621347173643235
ENTERTAINMENT : 0.9590432133589079
LIBRARIES_AND_DEMO : 0.9364774906916393
AUTO_AND_VEHICLES : 0.9251946293580051
HOUSE_AND_HOME : 0.8236488773552973
WEATHER : 0.8010831546880289
EVENTS : 0.7108202640189552
PARENTING : 0.6544059573507841
ART_AND_DESIGN : 0

The results seem different in case of Play store. There seem to be more apps on Google Play on practical purposes and not many apps made for fun. But if we explore the Play Store, we can see that the Family category mostly consists of Gaming apps for kids.
![Google Play Family apps](https://camo.githubusercontent.com/9bf24b9efc3d88a3d55f5c09e314987941f0bab5/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f64712d636f6e74656e742f3335302f7079316d385f66616d696c792e706e67)

But even then, the Play Store has better representation of practical apps than App Store. This could be confirmed from the frequency table for genres column of our data.

In [23]:
display_table(android_final, -4) # Genre

Tools : 8.450863138892023
Entertainment : 6.070179397495204
Education : 5.348076272142616
Business : 4.592124562789123
Productivity : 3.8925871601038025
Lifestyle : 3.8925871601038025
Finance : 3.7007785174320205
Medical : 3.5315355974275078
Sports : 3.463838429425702
Personalization : 3.317161232088458
Communication : 3.2381812027530184
Action : 3.102786866749408
Health & Fitness : 3.0802211440821394
Photography : 2.944826808078529
News & Magazines : 2.798149610741284
Social : 2.6627552747376737
Travel & Local : 2.324269434728647
Shopping : 2.245289405393208
Books & Reference : 2.1437436533904997
Simulation : 2.042197901387792
Dating : 1.8616721200496444
Arcade : 1.8503892587160102
Video Players & Editors : 1.771409229380571
Casual : 1.7601263680469368
Maps & Navigation : 1.399074805370642
Food & Drink : 1.241114746699763
Puzzle : 1.128286133363421
Racing : 0.9928917973598104
Role Playing : 0.9364774906916393
Libraries & Demo : 0.9364774906916393
Auto & Vehicles : 0.9251946293580051
S

So, here we can see that there are many apps on practical purposes. The difference between Category and Genre is not crystal clear, but we can tell that Genre is a granular version of Category, as it has more rows. So this is just the bigger picture of Category column. So, we will be using the Category column further.

Till now, we can say that the App Store is dominated by apps made for entertainment and fun, whereas the Google Play store has more of a balanced dominion by both entertainment and practical purpose apps.
Now, we will find the apps that attract users.

## Finding Popular apps by Genre on App Store
One way to find out what genres are the most popular is to calculate the average number of installs for each app genre. We can use the Installs column on the Play Store, but this information is missing in App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

Let's count average number of user ratings per app genre. This can be done in three steps:
  - Isolate the apps of each genre
  - Sum up the user ratings for the apps of that genre.
  - Divide the sum by the number of user ratings for each genre.

In [24]:
prime_genre = freq_table(ios_final, -5)

for genre in prime_genre:
    total = 0            #for storing number of ratings
    len_genre = 0        # for storing number of apps in a genre
    
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:
            user_ratings = float(app[5])
            total += user_ratings
            len_genre += 1
    avg_user_ratings = total / len_genre
    print(genre, avg_user_ratings)

Weather 52279.892857142855
Entertainment 14029.830708661417
Finance 31467.944444444445
Social Networking 71548.34905660378
Food & Drink 33333.92307692308
Music 57326.530303030304
Catalogs 4004.0
Shopping 26919.690476190477
Navigation 86090.33333333333
Health & Fitness 23298.015384615384
Reference 74942.11111111111
News 21248.023255813954
Business 7491.117647058823
Travel 28243.8
Utilities 18684.456790123455
Games 22788.6696905016
Education 7003.983050847458
Medical 612.0
Sports 23008.898550724636
Lifestyle 16485.764705882353
Productivity 21028.410714285714
Photo & Video 28441.54375
Book 39758.5


Looking at the average number of ratings for each genre, it seems that the users tend to install apps based on Navigation, Reference, Weather and Social Networking, over the other genres. The most popular genre is Navigation, with 86090 user ratings on an average.

But, looking at the data we can see that apps like Waze and Google Maps, which have close to half a million user reviews together;

In [25]:
for app in ios_final:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are.

Now, we should be removing such apps that skew the data as they increase the average too much, but we will leave this level of detail for now.

Right now a niche genre is Reference. Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [26]:
for app in ios_final:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


This genre seems to be more developable. as other genres are heavily skewed by some apps(eg: Music, Navigation), or require us to pay for external APIs(eg:Weather or News) or need an expert(eg: Banking) in that genre.

One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

Let's analyze data on Google Play now:
## Finding Popular apps by Genre on Play Store
We actually have the data on number of installs in the Play market, and can get a better picture on genre popularity.

However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [27]:
display_table(android_final, 5) # the Installs columns

1,000,000+ : 15.728308699086089
100,000+ : 11.55365000564143
10,000,000+ : 10.549475346947986
10,000+ : 10.199706645605326
1,000+ : 8.394448832223853
100+ : 6.916393997517771
5,000,000+ : 6.826131106848697
500,000+ : 5.562450637481666
50,000+ : 4.772650344127271
5,000+ : 4.513144533453684
10+ : 3.542818458761142
500+ : 3.2494640640866526
50,000,000+ : 2.3017037120613786
100,000,000+ : 2.1324607920568655
50+ : 1.9180864267178157
5+ : 0.7898002933543946
1+ : 0.5077287600135394
500,000,000+ : 0.270788672007221
1,000,000,000+ : 0.2256572266726842
0+ : 0.045131445334536835


But that's ok, because we dont want precise number of installs. We only want to find what app genres attracts more users.We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error.

We'll loop through the data and convert all the strings to corrected floats.

In [28]:
unique_genres = freq_table(android_final, 1)

for category in unique_genres:
    total = 0                #for storing sum of installs for a genre
    len_category = 0        #for storing the number of apps to each genre
    
    for app in android_final:
        category_app = app[1]
        if category_app == category:
            installs = app[5]
            installs = installs.replace(',', '')
            installs = installs.replace('+', '')
            installs = float(installs)
            total += installs
            len_category += 1
    avg_installs = total / len_category
    print(category, avg_installs)

BOOKS_AND_REFERENCE 8767811.894736841
FAMILY 3697848.1731343283
PRODUCTIVITY 16787331.344927534
FOOD_AND_DRINK 1924897.7363636363
EVENTS 253542.22222222222
AUTO_AND_VEHICLES 647317.8170731707
BUSINESS 1712290.1474201474
TOOLS 10801391.298666667
NEWS_AND_MAGAZINES 9549178.467741935
SPORTS 3638640.1428571427
SOCIAL 23253652.127118643
WEATHER 5074486.197183099
HEALTH_AND_FITNESS 4188821.9853479853
PHOTOGRAPHY 17840110.40229885
COMMUNICATION 38456119.167247385
PERSONALIZATION 5201482.6122448975
PARENTING 542603.6206896552
GAME 15588015.603248259
MAPS_AND_NAVIGATION 4056941.7741935486
TRAVEL_AND_LOCAL 13984077.710144928
DATING 854028.8303030303
ENTERTAINMENT 11640705.88235294
FINANCE 1387692.475609756
ART_AND_DESIGN 1986335.0877192982
COMICS 817657.2727272727
SHOPPING 7036877.311557789
EDUCATION 1833495.145631068
VIDEO_PLAYERS 24727872.452830188
HOUSE_AND_HOME 1331540.5616438356
LIFESTYLE 1437816.2687861272
MEDICAL 120550.61980830671
BEAUTY 513151.88679245283
LIBRARIES_AND_DEMO 638503.73493

If we analyze the data, we can make out that the most installs are for communication apps(38456119 installs). However, this number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [29]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

Removing such apps from our set will normalize the data and reduce the average by roughly ten times.

In [30]:
under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
print(sum(under_100_m) / len(under_100_m))

3603485.3884615386


We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player.

Also, the pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at thhe Books and References data set then;

In [31]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. But, looking at the data carefully, there's still a small number of extremely popular apps that skew the average:

In [32]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


Now, we can see that there are very few popular apps in this genre. Hence, if we make an app related to this genre, we may turn out to be profitable.

Let's try to get some app ideas for this, by printing a few apps in this genre;

In [33]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

There seem to be a lot of apps based on software and processing ebooks, collection of libraries and dictionaries,etc. So, it's probably not a good idea to develop a similar app, since there is a lot of competition.

But, we also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

We can see that there are a lot of libraries based on this. Hence, we definitely need to add some new features. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

And a good user interface, would also add up to make our app more popular.

# Conclusion
In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features, a good user interface, along with the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.